akka-projection-core/src/main/scala/akka/projection/scaladsl/SourceProvider.scala

```diff
@@ -33,3 +33,18 @@ trait VerifiableSourceProvider[Offset, Envelope] extends SourceProvider[Offset,
 }
 
 trait MergeableOffsetSourceProvider[Offset <: MergeableOffset[_], Envelope] extends SourceProvider[Offset, Envelope]
+
+trait QueryableForMaxOffsetSourceProvider[Offset, Envelope] extends SourceProvider[Offset, Envelope] {
+  def maxOffset(): Future[Option[Offset]]
```

Comment:
not sure I follow how we should implement this? is it a new interface that is implemented in the (r2dbc) query plugin to provide this capability?

That would then given an entityType and slice range, query the journal for the offset of latest event?

Reply:
For the event-sourced (and durable state also) source providers, yeah, I think this does require a new query interface for the read journals to implement (similar in projection for detecting if the source provider supports replaying).